Add Partial Update Support for AES-CBC-HMAC-SHA-ETM Ciphers

gowthamsk-arm opened the pull request:
This PR extends the functionality of the AES-CBC-HMAC-SHA-ETM provider-based ciphers, originally introduced in #22949 and discussed in issue #19932, by adding robust partial update support. This enables secure and efficient handling of fragmented or streaming data, which is critical for many real-world applications.

Summary of Changes:

1. Partial Update Support for ETM Ciphers

- Adds stateful handling for interleaved AES and HMAC contexts, allowing input to be processed in parts across multiple EVP_Update() calls.
- Uses cipher-specific update buffers (up to 128 bytes), superseding the default PROV_CIPHER_CTX buffer which is too small for SHA512-based interleaved modes.
- Implements custom padding and unpadding logic to correctly handle AES block-size padding while maintaining alignment with the larger SHA block size.

2. HMAC Calculation Optimizations

- Avoids redundant final HMAC tag computation during every update call by deferring final HMAC computation to EVP_Final() for the ETM ciphers.

3. Test Enhancements

- Enables partial update tests for AES-CBC-HMAC-SHA-ETM ciphers.
- Adds test vectors for AES-CBC-192-HMAC-SHAx-ETM variants.
- Introduces support for padding tests in the EVP test framework via Padding = ENABLED.

4. Other changes

- Adds a macro **EVP_MAX_BLOCK_LENGTH_CBC_HMAC_SHA_ETM** (128 bytes) to prevent misuse of **EVP_MAX_BLOCK_LENGTH** in buffer allocations.

gowthamsk-arm committed:
Add support for partial updates to AES-CBC-HMAC-SHA-ETM

This commit introduces robust support for partial updates in AES-CBC
HMAC-SHA-ETM provider-based ciphers, enabling secure and efficient
handling of streaming data with interleaved processing.

Key changes:

- Added support for saving intermediate SHA and AES states:
  Internal state saving was implemented to facilitate proper handling
  of partial update calls, enabling correct hash and cipher state
  restoration during multi-part operations.

- Configured cipher-specific partial update buffers:
  The generic PROV_CIPHER_CTX buffer (16 bytes) is insufficient for
  AES-CBC-HMAC-ETM ciphers, which use block sizes equal to the SHA
  digest block size (e.g., 64 bytes for SHA-512). This patch introduces
  custom per-cipher buffers sized accordingly and updates the
  aes_update and aes_final paths to use them.

- Added support for custom padding logic:
  Interleaved ciphers use larger internal block sizes but must still
  pad based on the AES block size (16 bytes). Custom padding and
  unpadding routines were added to handle this correctly without
  relying on EVP legacy logic.

- Enforced tag verification before unpadding:
  To prevent padding oracle attacks, the tag is now verified before
  any attempt is made to unpad and write decrypted plaintext.

- Enabled partial update tests for ETM ciphers:
  Updated test filters to include ciphers with
  EVP_CIPH_FLAG_ENC_THEN_MAC, enabling verification of AES-CBC HMAC-ETM
  implementations through partial update and fragmentation tests.

Signed-off-by: Jack Bond-Preston <jack.bond-preston@arm.com>
Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

gowthamsk-arm committed:
Add AES-192-HMAC-SHAx test vectors

This patch adds the missing test vectors for AES-192 HMAC-SHAx
interleaved ciphers.

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

gowthamsk-arm committed:
Add padding test setup for AES-CBC-HMAC-SHAx-ETM ciphers

The current EVP test infrastructure only validates ciphers with
padding disabled. This patch enhances the test framework to support
padding scenarios by introducing a new config option:
'Padding = ENABLED'.

Test vectors are added for AES-CBC-HMAC-SHA1/256/512-ETM ciphers
with padding enabled, verifying correct behavior of custom padding
and unpadding logic. This ensures better coverage for interleaved
cipher modes using block-aligned and misaligned plaintext inputs.

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

gowthamsk-arm committed:
Optimize HMAC calculation in interleaved AES-CBC+HMAC-SHA1

Currently for every EVP Update call, we calculate the final HMAC tag
which involves 2 hash calculations. This is redundant as the
subsequent Update calls will overwrite the tag. It is possible to
only calculate the final tag once at the end.

This patch optimizes this HMAC calculation. For Update calls, we
store the internal values of hash1 and return. Only when EVP final
call is made, we complete hash1 with padding and calculate hash2
(the final tag).

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

gowthamsk-arm committed:
Optimize HMAC calculation in interleaved AES-CBC+HMAC-SHA256

Currently for every EVP Update call, we calculate the final HMAC tag
which involves 2 hash calculations. This is redundant as the
subsequent Update calls will overwrite the tag. It is possible to
only calculate the final tag once at the end.

This patch optimizes this HMAC calculation. For Update calls, we
store the internal values of hash1 and return. Only when EVP final
call is made, we complete hash1 with padding and calculate hash2
(the final tag).

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

gowthamsk-arm committed:
Optimize HMAC calculation in interleaved AES-CBC+HMAC-SHA512

Currently for every EVP Update call, we calculate the final HMAC tag
which involves 2 hash calculations. This is redundant as the
subsequent Update calls will overwrite the tag. It is possible to
only calculate the final tag once at the end.

This patch optimizes this HMAC calculation. For Update calls, we
store the internal values of hash1 and return. Only when EVP final
call is made, we complete hash1 with padding and calculate hash2
(the final tag).

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

tom-cosgrove-arm reviewed (dismissed):
This change has been reviewed and approved internally

tom-cosgrove-arm reviewed (dismissed):
Reviewed and approved internally

gowthamsk-arm commented:
Hello OpenSSL maintainers,

We're seeking your input on a couple of design decisions related to this PR.

1. **New Macro EVP_MAX_BLOCK_LENGTH_CBC_HMAC_SHA_ETM for EVP_EncryptUpdate() calls**

In these ciphers, AES-CBC and HMAC-SHA are processed in an interleaved fashion (Encrypt-then-MAC), which means the EVP update calls will require more buffering than typical block ciphers.

Specifically, the effective block size is governed by the digest block size (e.g., 64 bytes for SHA-256, 128 bytes for SHA-512) — not the AES block size (16 bytes). The existing **EVP_MAX_BLOCK_LENGTH** (32 bytes) is therefore insufficient and potentially unsafe in this context.

To address this, we've proposed **EVP_MAX_BLOCK_LENGTH_CBC_HMAC_SHA_ETM** macro with strong inline documentation cautioning that:

- This macro is only for internal/static allocation for interleaved ciphers
- The preferred way to get buffer sizing is still EVP_CIPHER_get_block_size()
- This macro must not be used generically across cipher implementations

**Question:**
To prevent silent misuse or underallocation issues—particularly in cases where static buffers are used, we've introduced this macro as a conservative, mode-specific upper bound. Does this approach align with OpenSSL's conventions, or would you recommend a different way to handle buffer sizing during EVP update() in these interleaved ciphers?

2. **Reusing HMAC Key via SHA_CTX Snapshot**

To avoid requiring the HMAC key to be resupplied for each new hash computation, we snapshot the SHA context after initial key setup.

```
memcpy(&ctx->head, &ctx->head_init, sizeof(SHA_CTX));
memcpy(&ctx->tail, &ctx->tail_init, sizeof(SHA_CTX));
```

This allows us to quickly reinitialize the hash state without reprocessing the key material, which improves performance—particularly in scenarios involving repeated EVP_EncryptInit() / EVP_EncryptUpdate() cycles with the same key.

**Question:**
Is this memcpy-based reuse of SHA_CTX considered safe and supported within provider code?
Is there a preferred mechanism in OpenSSL for this kind of context preservation?

t8m commented:
@slontis @vdukhovni @levitte  might have some opinions I think.

vdukhovni commented:
With regard to `SHA_CTX`, there is no problem with the `base`, `default` and `fips` providers knowing the details of the structure layout because the `base` and `default` providers are built-in to lib crypto (have the same layout as the rest of the library), while the `fips` provider has its copies of the SHA routines independent of the runtime libcrypto.  So, if I'm not mistaken, either way, there is no opportunity for conflict.

This does not mean that other "third party" providers can do the same, but that should not matter in this PR.

gowthamsk-arm commented on include/openssl/evp.h:
```diff
@@ -37,6 +37,27 @@
 # define EVP_MAX_BLOCK_LENGTH            32
 # define EVP_MAX_AEAD_TAG_LENGTH         16
 
+/*
+ * Maximum block size for AES-CBC-HMAC-SHAx-ETM combined ciphers.
+ *
+ * These ciphers are designed by interleaving two cryptographic primitives:
+ * - AES in CBC mode for encryption/decryption
+ * - HMAC with SHA (e.g., SHA-256 or SHA-512) for authentication
+ *
+ * Due to the interleaved processing of both algorithms, the effective block
+ * size required for secure and correct operation is equal to the digest block
+ * size of the corresponding SHA algorithm (e.g., 64 bytes for SHA-256, 128 bytes
+ * for SHA-512), not the AES block size (16 bytes).
+ *
+ * WARNING: Do NOT use EVP_MAX_BLOCK_LENGTH (32 bytes) for buffer sizing with
+ * these ciphers.
+ *
+ * Instead, always use EVP_CIPHER_get_block_size() at runtime to determine
+ * the correct block size, or use EVP_MAX_BLOCK_LENGTH_CBC_HMAC_SHA_ETM macro for
+ * a safe upper bound when statically allocating buffers.
+ */
+# define EVP_MAX_BLOCK_LENGTH_CBC_HMAC_SHA_ETM 128
+
```
Yes that makes sense. My intention was to make it easier for users to size buffers safely when working with these interleaved ETM ciphers where block sizes are governed more by the digest size (e.g., 64 or 128 bytes) than AES's 16-byte block for update calls.

Is there a recommended way within the EVP framework or guidance we can provide in the docs to help users allocate appropriately sized buffers without relying on hardcoded assumptions? For example, should we emphasize EVP_CIPHER_get_block_size() as the runtime-safe approach, even if it's somewhat surprising for users who expect AES block size to be 16 bytes?

shahsb reviewed (commented):
Buffer Copies: Frequent small updates (<128B) may incur copy overhead.

Recommendation: Benchmark fragmented inputs (1B/16B/128B chunks) to quantify impact.

gowthamsk-arm commented:
Even with memcpy, this approach is still more optimal than recomputing or duplicating MAC/cipher state across updates. We are avoiding redundant work by preserving intermediate state, which outweighs the minor cost of the occasional small copy.

shahsb reviewed (commented):
**Important Security Implication:** During decryption, plaintext must not be output until the HMAC is verified in EVP_Final() (to prevent unauthenticated data exposure).

**Recommendation:** Explicitly document/verify that decrypted plaintext is buffered internally until HMAC validation.

shahsb reviewed (commented):
### **Edge Cases to Test**
1. **Fragmentation Extremes**:  
   - Single-byte updates.  
   - Inputs spanning SHA block boundaries (e.g., 127B → 1B).  
   - Empty final update (`EVP_Update(in=0)` followed by `EVP_Final()`).  
2. **Padding Boundaries**:  
   - Data ending exactly at AES block vs. requiring padding.  
   - Invalid padding after valid HMAC (decryption).  
3. **Key Sizes**: Validate AES-192-HMAC combinations (SHA256/SHA384/SHA512).

gowthamsk-arm commented:
> Important Security Implication: During decryption, plaintext must not be output until the HMAC is verified in EVP_Final() (to prevent unauthenticated data exposure).
> 
> Recommendation: Explicitly document/verify that decrypted plaintext is buffered internally until HMAC validation.

This has been considered in the implementation [here](https://github.com/openssl/openssl/pull/27609/commits/e3cd426cb2652ba25e6b9ce3557983a2c8577938#diff-ca389f84ec98a17b50cbdbb990609babca405bebc4b35c86e1771c0acf66515aR184). 

```
    if (!ctx->hw->cipher(ctx, buf, buf, ctx->bufsz)) {
        ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
        return 0;
    }

    if (CRYPTO_memcmp(sctx->exp_tag, sctx->tag, sctx->taglen) != 0)
        return 0;

    if (ctx->pad &&
        !ossl_interleaved_cipher_unpadblock(buf, &ctx->bufsz, blksz)) {
        /* ERR_raise already called */
        return 0;
    }

    if (outsize < ctx->bufsz) {
        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
        return 0;
    }

    memcpy(out, buf, ctx->bufsz);
    *outl = ctx->bufsz;
    ctx->bufsz = 0;
```
After the interleaved AES and HMAC operation, we first check the tag. Only if the verification succeeds, we proceed to copy the plaintext to the `out` buffer.

gowthamsk-arm commented:
For the test cases,
1. Fragmentation Extremes: The `cipher_test_enc()` function in `test/evp_test.c` exercises a range of input fragment sizes. For ETM-based ciphers, this PR extends test coverage by enabling fragmentation tests alongside existing test vectors. This helps validate partial update logic under varied input chunking scenarios.
2. Padding Boundaries: [This](https://github.com/openssl/openssl/pull/27609/commits/6a984ca3e72a33f7b0fa12baa51e329fceb22d45) commit introduces targeted tests for padding behavior, including vectors with data lengths that cross padding boundaries. These ensure correct handling of padding in both exact-block and partial-block conditions.
3. Key Sizes validate AES-192-HMAC: [This](https://github.com/openssl/openssl/pull/27609/commits/a1e162ec4b58d50557b0633a266a231d851f1c70) commit adds the AES-192 test vectors for different SHA variants.

t8m commented:
> **Important Security Implication:** During decryption, plaintext must not be output until the HMAC is verified in EVP_Final() (to prevent unauthenticated data exposure).
> 
> **Recommendation:** Explicitly document/verify that decrypted plaintext is buffered internally until HMAC validation.

This is bogus and it is an application-level concern, not for EVP API.

gowthamsk-arm commented:
@paulidale Thanks for the reply. I'll drop the patch. Would it be better to document the need to use EVP_CIPHER_get_block_size() for the ETM ciphers?

paulidale commented:
EVP_CIPHER_get_block_size() should be called to get the block size of any cipher.  The documentation only mentions `EVP_MAX_BLOCK_LENGTH` as the maximum block length of all ciphers.

I don't see a pressing need to spell out the need to call EVP_CIPHER_get_block_size(), but it wouldn't hurt.

paulidale commented:
We document `EVP_MAX_BLOCK_LENGTH` as being the largest cipher block size so it's use is not unreasonable.  It would be better to dynamically allocate I guess but that comes with penalties too.

We should update this if the maximum is larger.

gowthamsk-arm commented:
I'll add a patch on top updating the macro conditionally.

gowthamsk-arm committed:
Configure EVP_MAX_BLOCK_LENGTH macro based on ETM cipher

The interleaved AES-CBC-HMAC-SHAx-ETM ciphers support partial
updates and use a cipher block size equal to the corresponding
SHA compression block size (e.g., 64 bytes for SHA-256, 128
bytes for SHA-512). This exceeds the current EVP_MAX_BLOCK_LENGTH
value of 32 bytes.

As a result, EVP_Update() may overflow the block buffer when
processing ETM ciphers, especially with partial input.

This patch conditionally increases EVP_MAX_BLOCK_LENGTH to match the
maximum SHA block size when AES_CBC_HMAC_SHA_ETM_CAPABLE is defined,
preventing buffer overruns.

Non-ETM configurations retain the default 32-byte limit.

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

gowthamsk-arm commented:
The original motivation here was to replace the usage of Aarch64CryptoLib (unmaintained) in DPDK with OpenSSL (maintained) for these functions. It is not specifically for TLS. 
More info about the cipher can be found here https://github.com/openssl/openssl/issues/19932.

gowthamsk-arm commented:
We were working based on the DPDK spec which uses OpenSSL as PMD. DPDK uses this for high-throughput, low-latency packet processing. We have asked the author of https://github.com/openssl/openssl/issues/19932 for exact info on the spec. Due to time zone difference, it might take some time for his response.

gowthamsk-arm commented:
@mattcaswell Do you have any further input?

gowthamsk-arm commented:
Any further inputs on this?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

gowthamsk-arm commented:
On the V2 platform, optimizing the partial-update paths delivers ~62–64% performance uplift for SHA1/SHA256-based ETM ciphers and ~110% uplift for SHA512-based ETM cipher, measured on 8 KB buffers with small fragment sizes.

tom-cosgrove-arm reviewed (dismissed):
This has been reviewed and approved internally

jwright-arm commented:
Hi @t8m, the conflicts here are only due to CHANGES.md. Could you help us out with another review, or suggest someone that could help? Thanks!

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

gowthamsk-arm commented on include/openssl/evp.h:
```diff
@@ -34,7 +34,23 @@
 #define EVP_MAX_MD_SIZE 64 /* longest known is SHA512 */
 #define EVP_MAX_KEY_LENGTH 64
 #define EVP_MAX_IV_LENGTH 16
+
+/*
+ * EVP_MAX_BLOCK_LENGTH : static upper-bound for cipher blocks
+ *
+ * - Default (32): Two AES blocks are big enough for most ciphers.
+ *
+ * - ETM build (128): When AES-CBC is interleaved with HMAC in the
+ *   Encrypt-then-MAC ciphers (AES_CBC_HMAC_SHA_ETM_CAPABLE), the effective block
+ *   size is not 16B.  It inherits the digest-block size of the SHA variant in
+ *   use: SHA-256 = 64B, SHA-512 = 128B.
+ */
+#ifndef AES_CBC_HMAC_SHA_ETM_CAPABLE
 #define EVP_MAX_BLOCK_LENGTH 32
+#else
+#define EVP_MAX_BLOCK_LENGTH 128
+#endif
+
```
@t8m any further comments on this?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago
